File: tools/art/gen-sky.mjs
/**
 * Port 1:1 skryptu tools/art/gen-sky.py (PIL) na prymitywy pipeline'u – generuje sky-source.png 384x216
 * z paletą ze specyfikacji. Wynik ląduje w assets/raw/art/backgrounds/ i NADPISUJE proceduralne niebo (build-assets).
 * Uruchomienie: node tools/art/gen-sky.mjs
 */
import { create, save } from '../png.mjs';
import { rect, circle, hex } from '../pixel.mjs';

const WIDTH = 384, HEIGHT = 216, HORIZON_Y = 166;
const C_VIOLET = '#7a6aa6', C_PINK = '#d38f8e', C_AMBER = '#f0a874', C_GOLD = '#fbcc7c', C_SUN = '#ffe39a';
const C_HILLS = '#414b4e', C_CLOUD_DARK = '#a8645f', C_TOWER = '#8c8296';

const img = create(WIDTH, HEIGHT);
// PIL rectangle jest inkluzywne (x1, y1) – odwzorowujemy +1
const R = (x0, y0, x1, y1, c) => rect(img, x0, y0, x1 - x0 + 1, y1 - y0 + 1, c);

// 1. Gradient nieba – 4 pasma
for (const [y0, y1, c] of [[0, 45, C_VIOLET], [45, 90, C_PINK], [90, 135, C_AMBER], [135, HORIZON_Y, C_GOLD]]) R(0, y0, WIDTH, y1, c);

// 2. Słońce (elipsa poświaty + dysk)
const ellipse = (x0, y0, x1, y1, c) => { const cx = (x0 + x1) / 2, cy = (y0 + y1) / 2, rx = (x1 - x0) / 2, ry = (y1 - y0) / 2;
  for (let y = Math.floor(y0); y <= Math.ceil(y1); y++) for (let x = Math.floor(x0); x <= Math.ceil(x1); x++) if (((x - cx) / rx) ** 2 + ((y - cy) / ry) ** 2 <= 1) img.data.set(hex(c), (y * WIDTH + x) * 4); };
const sun_x = 240, sun_y = 148, sun_r = 16;
ellipse(sun_x - sun_r - 4, sun_y - sun_r - 2, sun_x + sun_r + 4, sun_y + sun_r + 2, C_AMBER);
circle(img, sun_x, sun_y, sun_r, C_SUN, true);

// 3. Płaskie chmury pikselowe
const cloud = (x, y, w, h) => { R(x, y, x + w, y + h - 2, C_GOLD); R(x + 2, y + h - 2, x + w - 2, y + h, C_CLOUD_DARK); };
cloud(40, 50, 60, 8); cloud(180, 75, 90, 10); cloud(310, 40, 50, 7);

// 4. Daleka wieża po lewej (<30 px) – trapez (45,166) (49,30) (55,30) (59,166)
for (let y = 30; y <= HORIZON_Y; y++) { const t = (y - 30) / (HORIZON_Y - 30); const xl = 49 - 4 * t, xr = 55 + 4 * t; R(Math.round(xl), y, Math.round(xr), y, C_TOWER); }

// 5. Wzgórza / ziemia od horyzontu w dół
R(0, HORIZON_Y, WIDTH, HEIGHT, C_HILLS);

save(img, 'assets/raw/art/backgrounds/sky-source.png');
console.log('Zapisano assets/raw/art/backgrounds/sky-source.png (384x216)');

File: tools/art/gen-sky.py
from PIL import Image, ImageDraw

WIDTH, HEIGHT = 384, 216
HORIZON_Y = 166

# Paleta ze specyfikacji
C_VIOLET = (122, 106, 166)  # #7a6aa6
C_PINK   = (211, 143, 142)  # #d38f8e
C_AMBER  = (240, 168, 116)  # #f0a874
C_GOLD   = (251, 204, 124)  # #fbcc7c
C_SUN    = (255, 227, 154)  # #ffe39a
C_HILLS  = (65, 75, 78)     # wzgórza
C_CLOUD_DARK = (168, 100, 95)
C_TOWER  = (140, 130, 150)

img = Image.new("RGB", (WIDTH, HEIGHT))
draw = ImageDraw.Draw(img)

# 1. Gradient nieba
bands = [
    (0, 45, C_VIOLET),
    (45, 90, C_PINK),
    (90, 135, C_AMBER),
    (135, HORIZON_Y, C_GOLD)
]
for y_start, y_end, color in bands:
    draw.rectangle([0, y_start, WIDTH, y_end], fill=color)

# 2. Słońce
sun_x, sun_y, sun_r = 240, 148, 16
draw.ellipse([sun_x - sun_r - 4, sun_y - sun_r - 2, sun_x + sun_r + 4, sun_y + sun_r + 2], fill=C_AMBER)
draw.ellipse([sun_x - sun_r, sun_y - sun_r, sun_x + sun_r, sun_y + sun_r], fill=C_SUN)

# 3. Płaskie chmury pikselowe
def draw_pixel_cloud(x, y, w, h):
    draw.rectangle([x, y, x + w, y + h - 2], fill=C_GOLD)
    draw.rectangle([x + 2, y + h - 2, x + w - 2, y + h], fill=C_CLOUD_DARK)

draw_pixel_cloud(40, 50, 60, 8)
draw_pixel_cloud(180, 75, 90, 10)
draw_pixel_cloud(310, 40, 50, 7)

# 4. Daleka wieża po lewej (<30 px)
draw.polygon([(45, HORIZON_Y), (49, 30), (55, 30), (59, HORIZON_Y)], fill=C_TOWER)

# 5. Wzgórza i linia horyzontu (y = 166)
draw.rectangle([0, HORIZON_Y, WIDTH, HEIGHT], fill=C_HILLS)

img.save("assets/raw/art/backgrounds/sky-source.png")
print("Zapisano sky-source.png (384x216)")
